Change env/service variables --> variables (#1987)
balena-io/docs: pages/learn/manage/variables.md

---
title: Variables

> excerpt: Set dynamic-named values on the dashboard that affect the way running processes will behave on the device by using variables.

Reviewer: ```suggestion
excerpt: Use variables to pass named values to the device which can be used to affect the way running apps and processes will behave.
```

---

# Variables

Variables allow you to provide runtime configuration to one or more running services without having to modify your source code. Variables can be added onto the device using the balenaCloud dashboard and can be assigned to either a specific service or all services running on the device. You can use variables to store secrets and other sensitive values out of your codebase and configure them when needed from the dashboard.
 

> Variables defined in the dashboard are added to the device as environement variables and can be accessed within the service that they were defined for as shown below:

Reviewer: ```suggestion
Variables defined in the dashboard are exposed on the device as environment variables and can be accessed within the service or services that they were defined for as shown below:
```

 
<img alt="Accessing variable" src="/img/variables/accessing_variables.png">

> __Note:__ Adding or modifying a variable for all services will restart all services on the device. Similarly for variables defined for specific services will only restart the affected service.

Reviewer: ```suggestion
__Note:__ Adding or modifying a variable for all services will restart all services on the device. Similarly, changing variables defined only for a specific service will only restart the affected service.
```

Values defined for individual devices always override those defined for the fleet. Values defined for specific services would override variables defined at the same level (Fleet or device level). So for any given variable, the **device service specific variable** will always have the top priority, followed by the **device variable for all services**, then the **fleet service specific variable**, and finally the **fleet variable for all services**.

__Note:__ Any variables defined through the dashboard do not apply to devices in [local mode][local-mode].

Values can be up to 1MB (or approximately 1 million characters) in size each. A device will re-download the variables every time the state changes in the API, which may potentially result in a lot of [network traffic][bandwidth-control]. The interval can be configured with the [BALENA_SUPERVISOR_POLL_INTERVAL configuration variable][poll-interval].

## Fleet-wide variables

Variables defined at the fleet level are available to all devices in that fleet unless they are redefined with a variable of the same name at the device level.

> Fleet-wide variables can be found from the Fleet Summary page by clicking the *Variables* tabs.

Reviewer: ```suggestion
Fleet-wide variables can be found from the Fleet Summary page by clicking the *Variables* tab.
```

To define a new variable, click the *Add variable* button in the upper-left corner.

<img alt="Add variable" src="/img/variables/add_application_variable.png" width="80%">

> In the dialog box that opens, select either a specific service or all services to apply the variable too.

Reviewer: ```suggestion
In the dialog box that opens, select either 'All services' or a specific service to which the variable will be applied.
```
